django/contrib/admin/views/autocomplete.py

```diff
@@ -41,11 +39,63 @@ def get_paginator(self, *args, **kwargs):
     def get_queryset(self):
         """Return queryset based on ModelAdmin.get_search_results()."""
         qs = self.model_admin.get_queryset(self.request)
+        qs = qs.complex_filter(self.source_field.get_limit_choices_to())
         qs, search_use_distinct = self.model_admin.get_search_results(self.request, qs, self.term)
         if search_use_distinct:
             qs = qs.distinct()
         return qs
 
+    def process_request(self, request):
+        """
+        Validate request integrity, extract and return request parameters.
+
+        Raises:
+            PermissionDenied: If the request is malicious or malformed.
+            Http404: If the target model admin is not configured properly.
+
+        """
+        # extract request parameters
+        term = request.GET.get('term', '')
+        try:
+            app_label = request.GET['app_label']
+            model_name = request.GET['model_name']
+            field_name = request.GET['field_name']
+        except KeyError as e:
+            raise PermissionDenied from e
+
+        # retrieve Django objects from parameters
+        try:
+            source_model = apps.get_model(app_label, model_name)
+        except LookupError as e:
+            raise PermissionDenied from e
+
+        try:
+            source_field = source_model._meta.get_field(field_name)
+        except FieldDoesNotExist as e:
+            raise PermissionDenied from e
+
+        try:
+            remote_model = source_field.remote_field.model
+        except AttributeError as e:
+            raise PermissionDenied from e
+        try:
+            model_admin = self.admin_site._registry[remote_model]
+        except KeyError as e:
+            raise PermissionDenied from e
```

Comment:
I think we should raise `BadRequest` instead of `PermissionDenied`, because missing parameter in `GET` has nothing to do with permissions. Maybe, I'm picky :shrug:

Reply:
The issue here is that we can't do the proper permission check (`if not self.has_perm(request) ...`) until after we've assigned `self.model_admin`, which we return from this method. 

_Maybe_ it's OK. But… it seems like we have the exact kind of _"Folks can probe the autocomplete view to find out the structure of the application"_ issue that we spend a lot of time thinking about and trying to avoid. 

If we just raise `PermissionDenied` for all of these cases we don't leak any info. Such exceptions are logged and a developer can use that to debug. Given that this is an ajax view, they're not going to see the debug output page anyway (without taking steps). 

I'm not convinced the different status code is worth the additional exposure (or even the effort justifying why it's OK). 

> 🤷 

Yes. 🙂

Reply:
This sounds like a non-obvious justification. Maybe it would be good to add a comment for future developers?